### ISSUE TITLE
Merge in Spark Dataset/Dataframe functionality from avro2spark

### ISSUE THREAD

**Issue by lossyrob** [locationtech/geotrellis#2051]
This issue to track the potential merging of the functionality found here: https://github.com/s22s/avro2spark into the GeoTrellis repository.

There's been a lot of really great work done, and I'm not privvy to the latest. What parts of that codebase would we merge in? What parts should be published as their own binaries? Is there a new subproject to be created?

TODO list from @metasim:

- [x] Implement our own schema conversion
- [ ] Decide if capability should be incorporated into GT primary, or stand-alone library. (If the latter, ideally  `AvroRecordCodec` would be in non-`spark` sub-project to minimize transitives.)
- [ ] Discuss requirements for user documentation
- [ ] (Code) Review [implementation](https://github.com/s22s/avro2spark/tree/master/src/main/scala/astraea/spark/avro) for organization and readability, correctness, etc.
- [ ] Determine which `Encoder`s to provide by default. Here are the ones we're currently using, (noting that our focus is on using `TileFeature` with a custom metadata container):
```scala
  implicit val mpdEncoder = AvroDerivedSparkEncoder[MODISProductDescriptor]
  implicit val tpeEncoder = AvroDerivedSparkEncoder[TemporalProjectedExtent]
  implicit val mbtEncoder = AvroDerivedSparkEncoder[MultibandTile]
  implicit val tfEncoder = AvroDerivedSparkEncoder[TileFeature[MultibandTile, MODISProductDescriptor]]
  implicit val tpeTfMbEncoder = AvroDerivedSparkEncoder[(TemporalProjectedExtent, TileFeature[MultibandTile, MODISProductDescriptor])]
  implicit val tpeTfSbEncoder = AvroDerivedSparkEncoder[(TemporalProjectedExtent, TileFeature[Tile, MODISProductDescriptor])]
```
- [ ] Determine requirements for GT-specific `org.apache.spark.ml.Transformer` implementations for converting Avro-encoded, DataFrame embedded structures into structures useful for SparkML. So far, the `spark2avro` code is only half the solution. To make the result useful to a data scientist, further transformations [such as these](https://gist.github.com/metasim/8c63bd53246b42004b9d4975e0b37a8d) are necessary (and the focus of my current work). This should be a point of further telecom discussion.

**Comment by pomadchin** [comment]
Looks like we'll get #1937 for free. For #1950 additional type classes needed to support all operations available on rdds on datasets.

**Comment by metasim** [comment]
One of the things outstanding in our `avro2spark` work is dropping the dependency on Databricks' `spark-avro` via our own Schema converter. I think that is in y'all's court (but let's negotiate if you need to).

As an aside, w.r.t. #1937, there's [this sub-project](https://github.com/apache/parquet-mr/tree/master/parquet-avro) of [`parquet-mr`](https://github.com/apache/parquet-mr) for converting Avro to Parquet.

**Comment by pomadchin** [comment]
@metasim @lossyrob I was talking about [this](http://spark.apache.org/docs/latest/sql-programming-guide.html#parquet-files)

**Comment by echeipesh** [comment]
@metasim That turned out to be pretty straight forward thing to do, the PR is outstanding and will come up today.

**Comment by metasim** [comment]
I'm not able to edit subject/descriptions, so here's a strawman todo list:

- [ ] Implement our own schema conversion
- [ ] Decide if capability should be incorporated into GT primary, or stand-alone library. (If the latter, ideally  `AvroRecordCodec` would be in non-`spark` sub-project to minimize transitives.)
- [ ] Discuss requirements for user documentation
- [ ] (Code) Review [implementation](https://github.com/s22s/avro2spark/tree/master/src/main/scala/astraea/spark/avro) for organization and readability, correctness, etc.
- [ ] Determine which `Encoder`s to provide by default. Here are the ones we're currently using, (noting that our focus is on using `TileFeature` with a custom metadata container):
```scala
  implicit val mpdEncoder = AvroDerivedSparkEncoder[MODISProductDescriptor]
  implicit val tpeEncoder = AvroDerivedSparkEncoder[TemporalProjectedExtent]
  implicit val mbtEncoder = AvroDerivedSparkEncoder[MultibandTile]
  implicit val tfEncoder = AvroDerivedSparkEncoder[TileFeature[MultibandTile, MODISProductDescriptor]]
  implicit val tpeTfMbEncoder = AvroDerivedSparkEncoder[(TemporalProjectedExtent, TileFeature[MultibandTile, MODISProductDescriptor])]
  implicit val tpeTfSbEncoder = AvroDerivedSparkEncoder[(TemporalProjectedExtent, TileFeature[Tile, MODISProductDescriptor])]
```
- [ ] Determine requirements for GT-specific `org.apache.spark.ml.Transformer` implementations for converting Avro-encoded, DataFrame embedded structures into structures useful for SparkML. So far, the `spark2avro` code is only half the solution. To make the result useful to a data scientist, further transformations [such as these](https://gist.github.com/metasim/8c63bd53246b42004b9d4975e0b37a8d) are necessary (and the focus of my current work). This should be a point of further telecom discussion.

**Comment by echeipesh** [comment]
I think this makes sense as part of core library under `spark-sql` package, to follow the split.

Some of the features that could appear in that package:
 - `Transformers` `Tile` and `Geometry` (converts geospatial types into ML features)
 - Spark SQL query interface into geotrellis layers
 - Spark SQL interface for Vector/Spatial join
 - Any Tile/Vector UDF that make sense in SQL context

**Comment by metasim** [comment]
@echeipesh The "Spark SQL interface for Vector/Spatial join" capability is what the CCRI guys just added to GeoMesa, and could be a potential point of synergy/overlap. Further,  "Spark SQL query interface into geotrellis layers" could likely be built directly upon a vector-oriented facility.

After @lossyrob gets off the road, let's set up a telecon so I can go over what I've learned on the first and last items. The last one has huge potential long term, but I think we'll have to go the UDT route like GeoMesa to overcome the nontrivial serialization cost (something I recently discovered) associated with UDFs and UDAFs and the need to go from Spark's `InternalRow` type and the outward facing `DataType` API values expected to/from UDFs and UDAFs.

cc: @anthonyccri @jnh5y
